src/detect-mqtt-protocol-version.c

```diff
@@ -0,0 +1,317 @@
+void DetectMQTTProtocolVersionFree(void *);
+
+static int DetectEngineInspectMQTTProtocolVersionGeneric(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id);
+
+typedef struct DetectMQTTProtocolVersionData_ {
+    uint8_t version;
+} DetectMQTTProtocolVersionData;
+
+/**
+ * \brief Registration function for mqtt.protocol_version: keyword
+ */
+void DetectMQTTProtocolVersionRegister (void)
+{
+    sigmatch_table[DETECT_AL_MQTT_PROTOCOL_VERSION].name = "mqtt.protocol_version";
+    sigmatch_table[DETECT_AL_MQTT_PROTOCOL_VERSION].desc = "match MQTT fixed header QOS level";
+    sigmatch_table[DETECT_AL_MQTT_PROTOCOL_VERSION].url = DOC_URL DOC_VERSION "/rules/mqtt-keywords.html#mqtt-protocol-version";
+    sigmatch_table[DETECT_AL_MQTT_PROTOCOL_VERSION].AppLayerTxMatch = DetectMQTTProtocolVersionMatch;
+    sigmatch_table[DETECT_AL_MQTT_PROTOCOL_VERSION].Setup = DetectMQTTProtocolVersionSetup;
+    sigmatch_table[DETECT_AL_MQTT_PROTOCOL_VERSION].Free  = DetectMQTTProtocolVersionFree;
+    sigmatch_table[DETECT_AL_MQTT_PROTOCOL_VERSION].RegisterTests = MQTTProtocolVersionRegisterTests;
+
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex);
+
+    DetectAppLayerInspectEngineRegister("mqtt.protocol_version",
+            ALPROTO_MQTT, SIG_FLAG_TOSERVER, 1,
+            DetectEngineInspectMQTTProtocolVersionGeneric);
+
+    mqtt_protocol_version_id = DetectBufferTypeGetByName("mqtt.protocol_version");
+}
+
+static int DetectEngineInspectMQTTProtocolVersionGeneric(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id)
+{
+    return DetectEngineInspectGenericList(tv, de_ctx, det_ctx, s, smd,
+                                          f, flags, alstate, txv, tx_id);
+}
+
+/**
+ * \internal
+ * \brief Function to match protocol version of an MQTT Tx
+ *
+ * \param det_ctx Pointer to the pattern matcher thread.
+ * \param f       Pointer to the current flow.
+ * \param flags   Flags.
+ * \param state   App layer state.
+ * \param txv     Pointer to the transaction.
+ * \param s       Pointer to the Signature.
+ * \param ctx     Pointer to the sigmatch that we will cast into DetectMQTTProtocolVersionData.
+ *
+ * \retval 0 no match.
+ * \retval 1 match.
+ */
+static int DetectMQTTProtocolVersionMatch(DetectEngineThreadCtx *det_ctx,
+                               Flow *f, uint8_t flags, void *state,
+                               void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+{
+    const DetectMQTTProtocolVersionData *de = (const DetectMQTTProtocolVersionData *)ctx;
+    uint8_t version;
+
+    if (!de)
+        return 0;
+
+    rs_mqtt_tx_get_protocol_version(state, &version);
```

Comment:
Would detect-engine-uint.h be useful here ?
If you want to specify in the signature version>3 or some similar stuff ?

Reply:
Makes sense.

Reply:
Done in v2, to be opened soon.

Reply:
Thanks